packages/usa-input-mask/src/index.js

```diff
@@ -1,124 +1,1095 @@
+      }
+
+      insertCharacterAtIndex(el, lastPosition, keyVal);
+
+      const HAS_MASK = getMaskInfo(inputEl.id, "HAS_MASK", false);
+      if (HAS_MASK) {
+        checkAndInsertMaskCharacters(el, lastPosition);
+      }
+    }
+
+    setCursorPosition(el, el.value.length);
+    handleValueChange(el);
+
+    event.preventDefault();
+    return false;
+  }
+
+  if (curPos < el.value.length) {
     if (
-      (matchesNumber && isInt) ||
-      (isCharsetPresent && matchesLetter && isLet)
-    ) {
-      newValue += strippedVal[charIndex];
-      charIndex += 1;
-    } else if (
-      (!isCharsetPresent && !isInt && matchesNumber) ||
-      (isCharsetPresent &&
-        ((matchesLetter && !isLet) || (matchesNumber && !isInt)))
+      FORMAT_CHARACTERS.indexOf(MASK[curPos]) > -1 ||
+      checkMaskForLetters(inputEl.id)
     ) {
-      return newValue;
-    } else {
-      newValue += placeholder[i];
+      setCursorPosition(inputEl, el.value.length);
+      event.preventDefault();
+      return false;
     }
-    // break if no characters left and the pattern is non-special character
-    if (strippedVal[charIndex] === undefined) {
-      break;
+  }
+
+  updateMaskMessage(
+    el,
+    statusMessageEl,
+    srStatusMessageEl,
+    keyCode,
+    event.key,
+    getCursorPosition(inputEl)
+  );
+
+  if (isValidCharacter(keyCode, MASK[getCursorPosition(el)])) {
+    if (keyCode >= KEYS.numberPadZero && keyCode <= KEYS.numberPadNine) {
+      keyCode -= 48;
+    }
+
+    let temp = event.key;
+
+    const FORCE_UPPER = getMaskInfo(inputEl.id, "FORCE_UPPER", false);
+
+    if (FORCE_UPPER) {
+      temp = temp.toUpperCase();
+    }
+
+    const FORCE_LOWER = getMaskInfo(inputEl.id, "FORCE_LOWER", false);
+
+    if (FORCE_LOWER) {
+      temp = temp.toLowerCase();
+    }
+
+    insertCharacterAtIndex(el, getCursorPosition(el), temp);
+
+    const HAS_MASK = getMaskInfo(inputEl.id, "HAS_MASK", false);
+    if (HAS_MASK) {
+      checkAndInsertMaskCharacters(el, getCursorPosition(el));
     }
   }
 
-  return newValue;
+  handleValueChange(el);
+
+  event.preventDefault();
+
+  return false;
 };
 
-const handleValueChange = (el) => {
-  const inputEl = el;
-  const id = inputEl.getAttribute("id");
-  inputEl.value = handleCurrentValue(inputEl);
+/**
+ * On init this function will create elements and update any
+ * attributes so it can validate user input.
+ *
+ * @param {HTMLInputElement} inputEl the components input
+ * @param {HTMLInputElement.dataset} options the components input attributes
+ */
+const enhanceInputMask = (inputEl) => {
+  const attrs = inputEl.attributes;
 
-  const maskVal = setValueOfMask(el);
-  const maskEl = document.getElementById(`${id}Mask`);
-  maskEl.textContent = "";
-  maskEl.replaceChildren(maskVal[0], maskVal[1]);
+  if (!attrs.mask) {
+    throw new Error(`${MASKED_INPUT_CLASS} is missing the mask attribute. Learn more at http://designsystem.digital.gov/components/input-mask/#available-attributes.`);
```

Comment:
This links to a table that will be created when https://github.com/uswds/uswds-site/pull/2073 is merged.